### PR TITLE
Implement `IsMinimalNonmonomial` for nonsolvable groups

### LINKED ISSUE #4452: IsMinimalNonmonomial only implemented for solvable groups
Reported by Bill Allombert via the GAP forum mailing list.

We have two options, taken from an Email by @ThomasBreuer  to the GAP forum:

- Implement a mehtod for nonsolvable groups. Just providing some simpleminded code just in order to get an answer instead of a "no method found" error message is not a good idea. If someone is really interested in the question for nonsolvable groups then probably the first step is to ask for minimal nonsolvable simple groups, and then one looks at the monomiality condition for the subgroups of these groups.

- Change the documentation such that it says that GAP can currently answer this question only for solvable groups and give a more informative error message if the argument is a nonsolvable group. An example for such a solution is the function `Omega`, which states: `At the moment methods exist only for abelian G and n=1.`

Option number 2 should be relatively easy to do, which is why I'm marking this
as good first issue.

### PR DESCRIPTION
**Pull request by jackschmidt**
Resolves #4452 

# Description

## Text for release notes 

Implement `IsMinimalNonmonomial` for nonsolvable groups

## Further details

The previous implementation followed van der Waall in only handling
solvable groups. Indeed in the paper,
[MR507832](https://mathscinet.ams.org/mathscinet-getitem?mr=507832)
van der Waall says the nonsolvable groups can easily be disposed
of using Thompson's N-groups papers, however he does not state the
result.

Luckily most of the minimal simple groups are also minimal
non-monomial. The exception is just PSL(3,3) which contains a subgroup
isomorphic to the non-monomial group SL(2,3). The groups of the form
PSL(2,q) are handled by Dickson's classification of their subgroups,
all of which are metabelian or non-solvable. Since the non-solvable
subgroups are eliminated by being minimal simple, we have only
monomial subgroups. The Suzuki groups require a short argument,
given in the source. Suzuki 1960 lists the maximal subgroups, and all
but the Sylow 2-normalizer are metabelian. The Sylow 2-normalizer's
characters are described in sufficient detail in the paper to ensure
they are monomial.

# Checklist for pull request reviewers

- [ ] proper formatting

If your code contains kernel C code, run `clang-format` on it; the 
simplest way is to use `git clang-format`, e.g. like this (don't
forget to commit the resulting changes):

    git clang-format $(git merge-base HEAD master)

- [ ] usage of relevant labels

   1. either `release notes: not needed` or `release notes: to be added`
   2. at least one of the labels `bug` or `enhancement` or `new feature`
   3. for changes meant to be backported to `stable-4.X` add the `backport-to-4.X` label
   4. consider adding any of the labels `build system`, `documentation`, `kernel`, `library`, `tests`

- [x] runnable tests
- [x] lines changed in commits are sufficiently covered by the tests
- [x] adequate pull request title
- [x] well formulated text for release notes
- [ ] relevant documentation updates
- [x] sensible comments in the code

### CODE REVIEW

File: lib/ctblmono.gi [hunk 1]
```diff
@@ -2176,6 +2176,74 @@ InstallMethod( IsMinimalNonmonomial,
...
+    # A non-solvable group all of whose proper quotients are monomial
+    # must have every proper quotient solvable. Hence G must be simple.
+    if not IsSimpleGroup(K) then
+      return false;
+    fi;
+
+    # If in addition, every proper subgroup is monomial, then every
+    # proper subgroup must be solvable, so G must be minimal simple.
+    #
+    # A minimal non-simple group must be (by Thompson's N-Groups paper):
```
Comment (fingolfin) [inline lib/ctblmono.gi:2199]: It is not hard to google it, but perhaps we could still insert e.g. the DO Ito make it absolutely clear?
```suggestion
    # A minimal non-simple group must be (by Corollary 1 in Thompson's 1968
    #  N-Groups paper, doi:10.1090/S0002-9904-1968-11953-6):
```

File: lib/ctblmono.gi [hunk 2]
```diff
@@ -2176,6 +2176,74 @@ InstallMethod( IsMinimalNonmonomial,
...
+    # * PSL(2,2^f) for f prime, and all such are minimal-non-M groups
+    # * PSL(2,3^f) for f an odd prime, and all such are minimal-non-M groups
+    # * PSL(3,3) which is NOT a minimal-non-M group (contains SL(2,3))
+    # * Sz(2^f) for f an odd prime, and all such are minimal-non-M groups
```
Comment (fingolfin) [inline lib/ctblmono.gi:2203]: That list seems to be missing the groups PSL(2,p) for p > 3 a prime congruent to 2 or 3 mod 5?

File: lib/ctblmono.gi [hunk 3]
```diff
@@ -2176,6 +2176,74 @@ InstallMethod( IsMinimalNonmonomial,
...
+    #
+    # The minimal-non-M part follows from the fact that metabelian groups
+    # are monomial groups.
+    # * For PSL(2,q) maximal subgroups are metabelian (since the non-solvable
+    #   maximals are ruled out by minimal simple)
```
Comment (fingolfin) [inline lib/ctblmono.gi:2208]: Maybe expand using the extra information from your excellent PR description? E.g. like this:

Dickson's classification of their subgroups,
all of which are metabelian or non-solvable
```suggestion
    # * For PSL(2,q) Dickson classified their subgroups; the maximal ones are
    #   either metabelian (hence monomial) or non-solvable; but this latter case
    #   is ruled out here by being minimal simple; hence these groups are minimal
    #   non-monomial.
```

File: lib/ctblmono.gi [hunk 4]
```diff
@@ -2176,6 +2176,74 @@ InstallMethod( IsMinimalNonmonomial,
...
+    # * For PSL(3,3) a subgroup isomorphic to SL(2,3) is not an M-group
+    # * For Sz(q), Suzuki 1960 shows the maximal subgroups are either the
+    #   the Sylow normalizer H, or one a few metabelian groups. The Sylow
+    #   normalizer's characters are calculated in section 11, page 126, and
+    #   are all induced from the monomial group Q, its Sylow 2-subgroup.
```
Comment (fingolfin) [inline lib/ctblmono.gi:2213]: "the the". Also, "the Sylow normalizer" makes it sound unique, but it isn't, right? And in the PR description you clarify that these are indeed normalizers of Sylow 2-groups, right? So perhaps
```suggestion
    # * For Sz(q), Suzuki 1960 shows the maximal subgroups are either the
    #   normalizer H of a Sylow 2-group Q, or one a few metabelian groups.
    #   The characters of H are calculated in section 11, page 126, and
    #   are all induced from the monomial group Q.
```

File: lib/ctblmono.gi [hunk 5]
```diff
@@ -2176,6 +2176,74 @@ InstallMethod( IsMinimalNonmonomial,
...
+    #   As an induced character can only be irreducible if the original
+    #   character was irreducible, and every irreducible of Q is induced
+    #   from a linear, and induction from that subgroup of Q and then to
+    #   H is the same as induction directly to H, we get that every
+    #   character of H is induced from a linear character (q-1 from H itself,
+    #   1 from Q, and 2 from a subgroup of Q).
```
Comment (fingolfin) [inline lib/ctblmono.gi:2219]:
```suggestion
    #   1 from Q, and 2 from a subgroup of Q). I.e., H is monomial.
```

File: lib/ctblmono.gi [hunk 6]
```diff
@@ -2176,6 +2176,74 @@ InstallMethod( IsMinimalNonmonomial,
...
+      else return false;
+      fi;
+    elif info.series = "L" then
+      if info.parameter[1] = 2 then
+        q := info.parameter[2];
+        p := SmallestRootInt(q);
+        f := LogInt(q,p);
+        if p = 2 and IsPrimeInt(f) then return true;
+        elif p = 3 and IsOddInt(f) and IsPrimeInt(f) then return true;
+        else return false;
```
Comment (fingolfin) [inline lib/ctblmono.gi:2233]:
```suggestion
        elif p = q and 0 = (p^2+1 mod 5) then return true;
        else return false;
```

File: lib/ctblmono.gi [hunk 7]
```diff
@@ -2176,6 +2176,74 @@ InstallMethod( IsMinimalNonmonomial,
...
+        fi;
+      else return false;
+      fi;
+    elif info.series = "2B" then
+      q := info.parameter;
+      f := LogInt(q,2);
+      if IsPrimeInt(f) then return true;
+      else return false;
```
Comment (fingolfin) [inline lib/ctblmono.gi:2241]: Matter of taste, but maybe:
```suggestion
      return IsPrimeInt(f);
```

File: lib/ctblmono.gi [hunk 8]
```diff
@@ -2176,6 +2176,74 @@ InstallMethod( IsMinimalNonmonomial,
...
+      fi;
+    else return false;
+    fi;
```
Comment (fingolfin) [inline lib/ctblmono.gi:2244]: I would do this, and then you can remove most of the `return false` above. But again: taste, I guess.
```suggestion
    fi;
    return false;
```

File: tst/teststandard/ctblmono.tst [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+gap> IsMinimalNonmonomial(SL(2,3));
+true
+gap> IsMinimalNonmonomial(GL(2,3));
+false
+gap> IsMinimalNonmonomial(PSL(2,4));
+true
+gap> IsMinimalNonmonomial(PGL(2,5));
+false
+gap> IsMinimalNonmonomial(PSL(2,7));
+false
```
Comment (fingolfin) [inline tst/teststandard/ctblmono.tst:16]: That's actually wrong (which agrees with Thompson's result and the other changes I suggested above. Indeed:
```
gap> all:=AllSubgroups(PSL(2,7));; Length(all); Number(all, IsMonomialGroup);
179
178
```

So we need this:
```suggestion
gap> IsMinimalNonmonomial(PSL(2,7));
true
```

File: tst/teststandard/ctblmono.tst [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+gap> IsMinimalNonmonomial(PSL(2,9));
+false
+gap> IsMinimalNonmonomial(PSL(2,11));
+false
+gap> IsMinimalNonmonomial(PSL(2,27));
+true
+gap> IsMinimalNonmonomial(PSL(3,3));
+false
+gap> IsMinimalNonmonomial(Sz(8));
+true
```
Comment (fingolfin) [inline tst/teststandard/ctblmono.tst:28]: For full code coverage, it would be nice to also test the negative case here
```suggestion
gap> IsMinimalNonmonomial(Sz(8));
true
gap> IsMinimalNonmonomial(Sz(32));
false
```

**Review by fingolfin — CHANGES_REQUESTED** [review]
Thank you very much for this very nice contribution! I left some remarks, and I think one case needs to be added.

File: tst/teststandard/ctblmono.tst (hunk 10 above)
Comment (jackschmidt) [reply tst/teststandard/ctblmono.tst:28]: I think Sz(2^9) is the first non-monomial (its only crime is that it contains Sz(2^3) as a subgroup). I'll check if GAP can handle such a large group.

Comment (jackschmidt) [reply tst/teststandard/ctblmono.tst:28]: I think this is now handled, though a bit awkwardly due to the size of the group. I'll submit a separate pull request that helps suzuki groups know they are not solvable.

File: lib/ctblmono.gi (hunk 2 above)
Comment (jackschmidt) [reply lib/ctblmono.gi:2208]: I've updated the comments now too. The code was already updated as per your suggestion

File: lib/ctblmono.gi [hunk 11]
```diff
@@ -2176,6 +2176,77 @@ InstallMethod( IsMinimalNonmonomial,
...
+    #
+    # The minimal-non-M part follows from the fact that metabelian groups
+    # are monomial groups.
+    # * For PSL(2,q) Dickson classified their subgroups; the maximal ones are
+    #   either metabelian (hence monomial) or non-solvable; but this latter case
+    #   is ruled out here by being minimal simple; hence these groups are minimal
+    #   non-monomial.
+    # * For PSL(3,3) a subgroup isomorphic to SL(2,3) is not an M-group
+    # * For Sz(q), Suzuki 1960 shows the maximal subgroups are either the
+    # * For Sz(q), Suzuki 1960 shows the maximal subgroups are either the
```
Comment (wilfwilson) [inline lib/ctblmono.gi:2218]: This line is duplicated:
```suggestion
    # * For Sz(q), Suzuki 1960 shows the maximal subgroups are either the
```

File: lib/ctblmono.gi [hunk 12]
```diff
@@ -2176,6 +2176,77 @@ InstallMethod( IsMinimalNonmonomial,
...
+    #   normalizer H of a Sylow 2-group Q, or one a few metabelian groups.
```
Comment (wilfwilson) [inline lib/ctblmono.gi:2219]:
```suggestion
    #   normalizer H of a Sylow 2-group Q, or one of a few metabelian groups.
```

**Review by wilfwilson — APPROVED** [review]
Thank you for this, especially the very helpful comments. I spotted just a couple of typos in them.

File: tst/teststandard/ctblmono.tst [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+# gap> for n in SizesPerfectGroups() do
+# > for k in [1..NrPerfectLibraryGroups(n)] do
+# > pg := PerfectGroup(IsPermGroup,n,k);
+# > Assert(0, IsMNMNaive(pg) = IsMinimalNonmonomial(pg) );
+# > od; od;
+# gap> for pg in PrimitiveGroupsIterator(NrMovedPoints,[1..20],IsSolvableGroup,false) do
+# > if IsNaturalAlternatingGroup(pg) or IsNaturalSymmetricGroup(pg) then continue; fi;
+# > if Size(pg) > 10^6 then continue; fi;
+# > Assert(0, IsMNMNaive(pg) = IsMinimalNonmonomial(pg) );
+# > od;
```
Comment (fingolfin) [inline tst/teststandard/ctblmono.tst:57]: Perhaps this could be made more efficient/concise by using `SimpleGroupsIterator`? Right now I think I suspect this code spends quite some time just checking perfect groups for simplicity.

that said, of course that's purely optional and just an idea you can ignore.

Comment (fingolfin) [reply tst/teststandard/ctblmono.tst:57]: I guess the goal is to verify the theorems by checking from first principles, and so you are going through some o the group libraries. OK, fair enough, ignore me then :-)

**Review by fingolfin — APPROVED** [review]
I've applied the corrections by @wilfwilson and squashed the four commits into one. If the tests pass, let's merge it. Thanks again, @jackschmidt !

File: tst/teststandard/ctblmono.tst (hunk 13 above)
Comment (jackschmidt) [reply tst/teststandard/ctblmono.tst:57]: Yes, sorry, just first principles to make sure nothing was omitted. I think Thompson's paper omitted a simple group (in a different result). I checked a bunch of solvable groups too, just to make sure nothing was amiss with the old solvable code. I think it completed up to 767. PrimitiveGroups got stuck on the maximal subgroups of M22 (I think M21 finished before I stopped the test.) I included them so future code reviewers (thank you!) would have a little easier time.
